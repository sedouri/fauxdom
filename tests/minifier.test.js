const {DOM, printHierarchy, describeNode} = require( "./utils" );
const htmlContent = require( "fs" ).readFileSync( "./tests/minifier.test.html", "utf8" );

describe( "Whitespace Minifier", () =>
{
	test( "Basic minification", () =>
	{
		const document = new DOM( htmlContent );
		document.minifyWhitespace();
		expect( document.innerHTML ).toMatchSnapshot();
	} );
	
	test( "Adjacent text nodes", () =>
	{
		const document = new DOM( htmlContent );
		document.querySelector( "span b" ).replaceWith( " and " );
		document.minifyWhitespace();
		expect( document.innerHTML ).toMatchSnapshot();
	} );
	
	test( "Custom inline elements", () =>
	{
		const document = new DOM( htmlContent );
		document.minifyWhitespace( ["div", "p", true] );
		expect( document.innerHTML ).toMatchSnapshot();
	} );
	
	test( "Minify inline CSS & JavaScript", () =>
	{
		const document = new DOM( htmlContent );
		document.minifyWhitespace( null, {
			inlineStyles( _node, value, _filePath )
			{
				return basicValueMinifier( value );
			},
			
			style( _node, value, _filePath )
			{
				return basicValueMinifier( value );
			},
			
			script( _node, value, filePath )
			{
				const val = basicValueMinifier( value );
				return (val ? `/*! ${filePath} */${val}` : null);
			}
		}, "./tests/minifier.test.html" );
		expect( document.innerHTML ).toMatchSnapshot();
	} );
} );

function basicValueMinifier( value )
{
	if ( value === true ) return null;
	
	const val = (value || "")
		.replace( /^\s+(.*?)\s*$/mg, "$1" )
		.replace( /\s*([(),])\s*/g, "$1" )
		.replace( /([^:]+:)\s*([^;]+(?:;|$))\s*/mg, "$1$2" )
		.replace( /[\n|\r|\n\r]+/g, "" )
		.replace( /;}/g, "}" )
		.replace( /;$/, "" );
	
	return (val === "" ? null : val);
}