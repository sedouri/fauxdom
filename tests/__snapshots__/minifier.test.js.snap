// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Whitespace Minifier Adjacent text nodes 1`] = `
"<!DOCTYPE html><html><head><title>This is a title</title><style>
		body
		{
			padding: 1em;
			background: #eee;
		}
		
		.copyright span
		{
			white-space: nowrap;
		}
	</style><style>
		
	</style><script src=\\"script.js\\" type=\\"text/javascript\\"></script><script>
		document.addEventListener( \\"DOMContentLoaded\\", () =>
		{
			console.log( {
				key: \\"value\\"
			} );
		} );
	</script><script>
		
	</script></head><body><div style>div content</div>text <a href=\\"#\\">link content</a> <span style=\\"padding: 1px 4px; background: yellow;\\">more text</span> <span>text</span> <!-- Comment --><span>text and text</span><!-- Comment --><div>text <b>text</b></div><!-- Comment --><div>text <b>text</b> j</div><pre>
		Don't minify
	</pre><content><h1>Header</h1><p>Paragraph 1</p><p>See the <a href=\\"/privacy.html\\">Privacy Policy</a> for more details.</p><hr><p>Email <a href=\\"mailto:here@there.com\\">here@there.com</a>, message <a href=\\"https://service.com/\\">@name</a> on Service.</p></content><footer><span class=\\"copyright\\"><span>Copyright &copy; Name.</span> <span>All rights reserved.</span></span></footer></body></html>"
`;

exports[`Whitespace Minifier Basic minification 1`] = `
"<!DOCTYPE html><html><head><title>This is a title</title><style>
		body
		{
			padding: 1em;
			background: #eee;
		}
		
		.copyright span
		{
			white-space: nowrap;
		}
	</style><style>
		
	</style><script src=\\"script.js\\" type=\\"text/javascript\\"></script><script>
		document.addEventListener( \\"DOMContentLoaded\\", () =>
		{
			console.log( {
				key: \\"value\\"
			} );
		} );
	</script><script>
		
	</script></head><body><div style>div content</div>text <a href=\\"#\\">link content</a> <span style=\\"padding: 1px 4px; background: yellow;\\">more text</span> <span>text</span> <!-- Comment --><span>text <b>text</b> text</span><!-- Comment --><div>text <b>text</b></div><!-- Comment --><div>text <b>text</b> j</div><pre>
		Don't minify
	</pre><content><h1>Header</h1><p>Paragraph 1</p><p>See the <a href=\\"/privacy.html\\">Privacy Policy</a> for more details.</p><hr><p>Email <a href=\\"mailto:here@there.com\\">here@there.com</a>, message <a href=\\"https://service.com/\\">@name</a> on Service.</p></content><footer><span class=\\"copyright\\"><span>Copyright &copy; Name.</span> <span>All rights reserved.</span></span></footer></body></html>"
`;

exports[`Whitespace Minifier Custom inline elements 1`] = `
"<!DOCTYPE html><html><head><title>This is a title</title><style>
		body
		{
			padding: 1em;
			background: #eee;
		}
		
		.copyright span
		{
			white-space: nowrap;
		}
	</style><style>
		
	</style><script src=\\"script.js\\" type=\\"text/javascript\\"></script><script>
		document.addEventListener( \\"DOMContentLoaded\\", () =>
		{
			console.log( {
				key: \\"value\\"
			} );
		} );
	</script><script>
		
	</script></head><body><div style>div content</div> text <a href=\\"#\\">link content</a> <span style=\\"padding: 1px 4px; background: yellow;\\">more text</span> <span>text</span> <!-- Comment --><span>text <b>text</b> text</span> <!-- Comment --><div>text <b>text</b></div> <!-- Comment --><div>text <b>text</b> j</div><pre>
		Don't minify
	</pre><content><h1>Header</h1><p>Paragraph 1</p> <p>See the <a href=\\"/privacy.html\\">Privacy Policy</a> for more details.</p><hr><p>Email <a href=\\"mailto:here@there.com\\">here@there.com</a>, message <a href=\\"https://service.com/\\">@name</a> on Service.</p></content><footer><span class=\\"copyright\\"><span>Copyright &copy; Name.</span> <span>All rights reserved.</span></span></footer></body></html>"
`;

exports[`Whitespace Minifier Minify inline CSS & JavaScript 1`] = `
"<!DOCTYPE html><html><head><title>This is a title</title><style>body{padding:1em;background:#eee}.copyright span{white-space:nowrap}</style><script src=\\"script.js\\" type=\\"text/javascript\\"></script><script>/*! ./tests/minifier.test.html */document.addEventListener(\\"DOMContentLoaded\\",()=>{console.log({key:\\"value\\"})})</script></head><body><div>div content</div>text <a href=\\"#\\">link content</a> <span style=\\"padding:1px 4px;background:yellow\\">more text</span> <span>text</span> <!-- Comment --><span>text <b>text</b> text</span><!-- Comment --><div>text <b>text</b></div><!-- Comment --><div>text <b>text</b> j</div><pre>
		Don't minify
	</pre><content><h1>Header</h1><p>Paragraph 1</p><p>See the <a href=\\"/privacy.html\\">Privacy Policy</a> for more details.</p><hr><p>Email <a href=\\"mailto:here@there.com\\">here@there.com</a>, message <a href=\\"https://service.com/\\">@name</a> on Service.</p></content><footer><span class=\\"copyright\\"><span>Copyright &copy; Name.</span> <span>All rights reserved.</span></span></footer></body></html>"
`;
